src/main.c

```diff
@@ -351,18 +362,120 @@ key_binding_func_key (MBWindowManager   *wm,
   hd_dbus_send_event (s);
 }
 
+/* Close all input devices previously opened */
+static void
+close_input_devices (Display *dpy)
+{
+  int i;
+
+  for (i = 0; i < xi_devices->len; i++)
+    {
+      hd_xi_device *xi_dev = &g_array_index (xi_devices, hd_xi_device, i);
+
+      if (xi_dev->dev)
+        {
+          XCloseDevice (dpy, xi_dev->dev);
+          xi_dev->dev = NULL;
+        }
+    }
+}
+
+/* Fill in the array with input devices type is touchscreen */
+static void
+enumerate_input_devices (Display *dpy)
+{
+  static Atom atom_touchscreen = None;
+  XDeviceInfo *devinfo;
+  int i, ndev;
+  GArray *eclass;
+
+  if (xi_devices)
+    {
+      close_input_devices (dpy);
+      g_array_free (xi_devices, TRUE);
+      xi_devices = NULL;
+    }
+
+  if (atom_touchscreen == None)
+    atom_touchscreen = XInternAtom (dpy, XI_TOUCHSCREEN, True);
+
+  if (atom_touchscreen == None)
+    return;
```

Comment:
May be replaced with:
```
  if (atom_touchscreen == None) {
    atom_touchscreen = XInternAtom (dpy, XI_TOUCHSCREEN, True);
    if (atom_touchscreen == None)
      return;
  }
```

If we have `atom_touchscreen != None` from the beginning, we will have two equal `if (...)` statements both of which will be false.

Reply:
I don;t really like such nested ifs, that's why I do it like that. Regarding the performance - I can bet compiler will optimize.

Reply:
> Regarding the performance - I can bet compiler will optimize.

According to my tests with `gcc -O2 -S`, it generates additional assembler instruction for the case without nested if's (on amd64 arch). When gcc doesn't use -O2, the instruction set size is the same, but it executes the check inside the if statement two times, just like it's done in the code.

Anyway, if you don't like nested if's so much, let's leave it this way.